frontend/src/components/common/Resource/PortForward.tsx

```diff
@@ -1,78 +1,167 @@
 }
 
 export default function PortForward(props: PortForwardProps) {
-  const { isPod, containerPort, namespace, name, isPodRunning } = props;
+  const { containerPort, resource } = props;
+  const isPod = resource?.metadata?.kind !== 'Service';
+  const service = !isPod ? (resource as Service) : undefined;
+  const namespace = resource?.metadata?.namespace;
+  const name = resource?.metadata?.name;
   const [isCurrentPodOrServicePortForwarding, setIsCurrentPodOrServicePortForwarding] =
     React.useState(false);
   const [error, setError] = React.useState(null);
   const [portForward, setPortForward] = React.useState<PortForwardState | null>(null);
   const [loading, setLoading] = React.useState(false);
   const cluster = getCluster();
-  const { t } = useTranslation('frequent');
+  const { t } = useTranslation(['frequent', 'resource']);
+  const [pods, podsFetchError] = Pod.useList({ labelSelector: getPodsSelectorFilter(service) });
+  const massagedContainerPort =
+    typeof containerPort === 'string' && isNaN(parseInt(containerPort))
+      ? getPortNumberFromPortName(pods[0].spec.containers, containerPort)
+      : containerPort;
+
+  React.useEffect(() => {
+    if (!cluster) {
+      return;
+    }
+    listPortForward(cluster).then(result => {
+      const portForwards = result;
+      if (!portForwards || portForwards.length === 0) {
+        return;
+      }
+      for (const item of portForwards) {
+        if (
+          (item.namespace === namespace || item.serviceNamespace === namespace) &&
+          (item.pod === name || item.service === name) &&
+          item.cluster === cluster &&
+          item.targetPort === massagedContainerPort.toString()
+        ) {
+          setIsCurrentPodOrServicePortForwarding(true);
+          setPortForward(item);
+          return;
+        }
+      }
+      const massagedPortForwards = [...portForwards];
+      const portForwardsInStorage = localStorage.getItem(PORT_FORWARDS_STORAGE_KEY);
+      const parsedPortForwards = JSON.parse(portForwardsInStorage || '[]');
+      parsedPortForwards.forEach((portforward: any) => {
+        const isPortForwardInStorage = portForwards.find((pf: any) => pf.id === portforward.id);
+        if (!isPortForwardInStorage) {
+          portforward.status = PORT_FORWARD_STOP_STATUS;
+          massagedPortForwards.push(portforward);
+        }
+      });
+      localStorage.setItem(PORT_FORWARDS_STORAGE_KEY, JSON.stringify(massagedPortForwards));
+    });
+  }, []);
 
   if (!helpers.isElectron()) {
     return null;
   }
 
+  if (!isPod && podsFetchError) {
+    return null;
+  }
+
+  if (!isPod && (!pods || pods.length === 0)) {
+    return null;
+  }
+
   function handlePortForward() {
-    if (!namespace || !containerPort || !cluster) {
+    if (!namespace || !cluster) {
       return;
     }
 
     setError(null);
 
     const massagedResourceName = name || '';
+    const massagedResourceNamespace = isPod ? namespace : pods[0].metadata.namespace;
```

Comment:
I believe this can be called `podNamespace`

Reply:
fixed